path_dynamic_creds.go

```diff
@@ -0,0 +1,275 @@
+package openldap
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-ldap/ldif"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/vault-plugin-secrets-openldap/client"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/template"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/mitchellh/mapstructure"
+	"golang.org/x/text/encoding/unicode"
+)
+
+func (b *backend) pathDynamicCredsRead(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("name").(string)
+
+	// Get the role and LDAP configs
+	dRole, err := retrieveDynamicRole(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve dynamic role: %w", err)
+	}
+	if dRole == nil {
+		return nil, nil
+	}
+
+	config, err := readConfig(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if config == nil {
+		return nil, fmt.Errorf("missing OpenLDAP configuration")
+	}
+
+	// Generate dynamic data
+	username, err := generateUsername(req, dRole)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate username: %w", err)
+	}
+	password, err := b.GeneratePassword(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Apply the template & execute
+	now := time.Now()
+	exp := now.Add(dRole.DefaultTTL)
+	templateData := dynamicTemplateData{
+		Username:              username,
+		Password:              password,
+		DisplayName:           req.DisplayName,
+		RoleName:              roleName,
+		IssueTime:             now.Format(time.RFC3339),
+		IssueTimeSeconds:      now.Unix(),
+		ExpirationTime:        exp.Format(time.RFC3339),
+		ExpirationTimeSeconds: exp.Unix(),
+	}
+	dns, err := b.executeLDIF(config.LDAP, dRole.CreationLDIF, templateData, false)
+	if err != nil {
+		// Creation failed, attempt a rollback if one is specified
+		if dRole.RollbackLDIF == "" {
+			return nil, fmt.Errorf("failed to create user: %w", err)
+		}
+
+		merr := multierror.Append(fmt.Errorf("failed to create user: %w", err))
+		_, err = b.executeLDIF(config.LDAP, dRole.RollbackLDIF, templateData, true)
+		if err != nil {
+			merr = multierror.Append(fmt.Errorf("failed to roll back user creation: %w", err))
+		}
+		return nil, merr
+	}
+	respData := map[string]interface{}{
+		"username":            username,
+		"password":            password,
+		"distinguished_names": dns,
+	}
+	internal := map[string]interface{}{
+		"name": roleName,
+		// Including the deletion_ldif in the event that the role is deleted while
+		// leases are active otherwise leases will fail to revoke
+		"deletion_ldif": dRole.DeletionLDIF,
+		"template_data": templateData,
+	}
+	resp := b.Secret(secretCredsType).Response(respData, internal)
+	resp.Secret.TTL = dRole.DefaultTTL
+	resp.Secret.MaxTTL = dRole.MaxTTL
+
+	return resp, nil
+}
+
+func encodeUTF16LE(str string) (string, error) {
```

Comment:
Nit: can we move this to be on L243 or so, so that it's closer to where it's being called?

Reply:
Done